A Launch script with Best Recipe of Deep Learning on Intel Xeon CPU

jingxu10 opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/63556

Usage: `python -m torch.backends.xeon.launch [--knobs] <script> [script parameters]`

jingxu10 commented:
Hi @VitalyFedyunin , any updates?

dzhulgakov commented:
Given that the script is cpu specific atm, should it be called `launch_cpu` or something like that?

@kiukchung might also want to take a look. It might not be a bad idea to embed it in to `torchrun` command that is a standard way for launching multi-worker jobs: https://pytorch.org/docs/stable/elastic/run.html That way UX can be similar for cpu and gpu

jingxu10 commented:
@dzhulgakov It is a fantastic idea to merge this script into `torchrun`.

jingxu10 commented:
Hi @dzhulgakov , do you have updates about embedding this PR to `torchserve`? `torchserve` is a wrap script which imports `torch.distributed.run`. Should we do similar work to import the PRed launch module in `torchserve` with a flag to choose from or integrate the functionality into `torch.distributed.run`?

jingxu10 commented:
Could you remove the stale label?

malfet commented on torch/utils/launch.py:
```diff
@@ -0,0 +1,805 @@
+    def __init__(self):
+
+        self.cpuinfo = []
+        if platform.system() == "Windows":
+            raise RuntimeError("Windows platform is not supported!!!")
```
Does it mean that scripts supports MacOS platform?

jingxu10 replied:
Probably doesn't work on macos neither, will double check with dev team.

malfet commented on torch/utils/launch.py:
```diff
@@ -0,0 +1,805 @@
+    args = parse_args()
+    if args.log_path:
+        path = os.path.dirname(args.log_path if args.log_path.endswith('/') else args.log_path + '/')
+        if not os.path.exists(path):
+            os.makedirs(path)
```
Please use `exist_ok` flag, see https://docs.python.org/3.7/library/os.html#os.makedirs
```suggestion
        os.makedirs(path, exist_ok=True)
```

jingxu10 replied:
Sure.

malfet commented on torch/utils/launch.py:
```diff
@@ -0,0 +1,805 @@
+            if len(matches) > 0:
+                lst_valid.append(item)
+            else:
+                logger.warning("{} doesn't exist. Removing it from LD_PRELOAD.".format(item))
```
We are in Python-3 world, please use f-strings
```suggestion
                logger.warning(f"{item} doesn't exist. Removing it from LD_PRELOAD.")
```

jingxu10 replied:
Sure.

malfet commented on torch/utils/launch.py:
```diff
@@ -0,0 +1,805 @@
+            for ip in ip_list[1:]:
+                completed_process = subprocess.run("ssh -o PasswordAuthentication=no {} ':'".format(ip), shell=True)
+                if completed_process.returncode != 0:
+                    logger.error("Passwordless SSH login to {} failed, please make sure you have setup SSH public key right")
```
Doesn't it miss an argument?
```suggestion
                    logger.error(f"Passwordless SSH login to {ip} failed, please make sure you have setup SSH public key right")
```

malfet commented on torch/utils/launch.py:
```diff
@@ -0,0 +1,805 @@
+            logger.info("Begin to validate the ip connect")
+            args.master_addr = ip_list[0]
+            for ip in ip_list[1:]:
+                completed_process = subprocess.run("ssh -o PasswordAuthentication=no {} ':'".format(ip), shell=True)
```
Please use f-strings
```suggestion
                completed_process = subprocess.run(f"ssh -o PasswordAuthentication=no {ip} ':'", shell=True)
```

malfet commented on torch/utils/launch.py:
```diff
@@ -0,0 +1,805 @@
+                    logger.error("Please make sure ninstances * ncore_per_instance <= total_cores")
+                    exit(-1)
+            if args.latency_mode:
+                logger.warning('--latency_mode is exclusive to --ninstances, --ncore_per_instance, --node_id and --use_logical_core. They won\'t take effect even they are set explicitly.')
```
Please use double quotes consistently throughout the code
```suggestion
                logger.warning("--latency_mode is exclusive to --ninstances, --ncore_per_instance, --node_id and --use_logical_core. They won\'t take effect even they are set explicitly.")
```

malfet reviewed (changes requested):
Please explicitly call out whether MacOS is supported or not
Use f-strings rather than `.format` and please use double quotes per https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#strings

kiukchung commented on torch/utils/launch_cpu_inference.py:
```diff
@@ -0,0 +1,577 @@
+                            or /.local/lib/ or /usr/local/lib/ or /usr/local/lib64/ or /usr/lib or /usr/lib64 or
+                           {expanduser("~")}/.local/lib/ so the LD_PRELOAD environment variable will not be set. This may drop the performance""")
+
+    def logger_env(self, env_name=""):
```
this method logs the environment variable name and value correct? then it should be called `log_env_var(env_var_name)`

jingxu10 replied:
OK. This prints an environment variable with name and value pair.

kiukchung commented on torch/utils/launch_cpu_inference.py:
```diff
@@ -0,0 +1,577 @@
+        if env_name in os.environ:
+            logger.info(f"{env_name}={os.environ[env_name]}")
+
+    def set_env(self, env_name, env_value=None):
+        if not env_value:
+            logger.warning(f"{env_name} is None")
```
was the default value of `None` for `env_value` intentional? if it was intentional (to clear env vars) then a `warning` is a bit misleading (change to info). Also the message should be clarified as:

```
logger.info(f"Clearing environment variable: {env_name}. Previous value: {previous_value}")
```

jingxu10 replied:
I'll remove this default value of `env_value` parameter.

kiukchung commented on torch/utils/launch_cpu_inference.py:
```diff
@@ -0,0 +1,577 @@
+        if env_name not in os.environ:
+            os.environ[env_name] = env_value
+        elif os.environ[env_name] != env_value:
+            logger.warning(f"{env_name} in environment variable is {os.environ[env_name]} while the value you set is {env_value}")
```
same here, if it is a valid action to set env vars that are already present this should be an "info" level log. Also the message can be improved:

```
logger.warning(f"Overriding already set environment variable: {env_name}. New value: {env_value}. Previous value: {prev_value}"
```

jingxu10 replied:
We don't mean to override the environment variable. We take values that are set by environment variables as the highest priority. This just notifies the user that he/she has set values via the environment variable, which is different than that set in the script.

kiukchung reviewed (commented):
Thank you for working on this.

Please include unittests with sufficient and necessary test-cases to assert the correctness of the launcher. Feel free to mock certain system-dependent calls where appropriate.

jingxu10 commented on setup.py:
```diff
@@ -874,7 +874,7 @@ def make_relative_rpath_args(path):
             'convert-caffe2-to-onnx = caffe2.python.onnx.bin.conversion:caffe2_to_onnx',
             'convert-onnx-to-caffe2 = caffe2.python.onnx.bin.conversion:onnx_to_caffe2',
-            'torchrun = torch.distributed.run:main',
+            'torchrun = torch.utils.launch:main',
```
`torch.utils.launch_cpu_inference` -> `torch.utils.inference.run`
Should I create a directory `inference` inside `torch/utils`, and move the `torch/utils/launch_cpu_inference.py` to `torch/utils/inference/run.py`?

jingxu10 commented on torch/utils/launch.py:
```diff
@@ -0,0 +1,29 @@
+    parser = argparse.ArgumentParser()
+    group = parser.add_mutually_exclusive_group(required=True)
+    group.add_argument('--inference', default=False, action="store_true")
+    group.add_argument('--distributed', default=False, action="store_true")
```
Sure.

jingxu10 replied:
Hi @kiukchung , do you expect to use it as the following?
torchrun xxxxx => DDP
torchrun --inference xxxxxx => inference

jingxu10 replied:
Since we introduce inference mode into torchrun, I think it makes sense to ask users to choose whether to use it in inference mode or ddp mode.

jingxu10 commented on torch/utils/launch_cpu_inference.py:
```diff
@@ -0,0 +1,577 @@
+        if platform.system() in ["Windows", "Darwin"]:
+            raise RuntimeError(f"{platform.system()} is not supported!!!")
+        elif platform.system() == "Linux":
+            args = ["lscpu", "--parse=CPU,Core,Socket,Node"]
```
On CentOS 7, `lscpu` doesn't take the `--json` argument.
```
# lscpu -e=CPU,Core,Socket,Node --json
lscpu: unrecognized option '--json'
Try 'lscpu --help' for more information.
# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
...
```

jingxu10 commented on torch/utils/launch_cpu_inference.py:
```diff
@@ -0,0 +1,577 @@
+    # multi-instance control
+    group.add_argument("--ncore_per_instance", metavar="\b", default=-1, type=int,
+                       help="Cores per instance")
+    group.add_argument("--ninstances", metavar="\b", default=-1, type=int,
```
This script is supposed for inference only at this moment. Both of these 2 arguments handle inference with multiple independent instances on a single node. An instance maps to a process.
`--ncore_per_instance` indicates how many cores are bound to an instance. How about making `--ninstances` to `--nprocs`?

jingxu10 commented on torch/utils/launch_cpu_inference.py:
```diff
@@ -0,0 +1,577 @@
+from argparse import ArgumentParser, REMAINDER
+from argparse import RawTextHelpFormatter
+import logging
+import psutil
```
Removed.

jingxu10 commented on torch/utils/launch_cpu_inference.py:
```diff
@@ -0,0 +1,577 @@
+        self.logical_core_socket_map = {}   # logical core to numa node id
+
+        self.nodes = int(max([line[3] for line in self.cpuinfo])) + 1
+        self.node_physical_cores = []  # node_id is index
+        self.node_logical_cores = []   # node_id is index
+        self.physical_core_node_map = {}  # phyical core to numa node id
+        self.logical_core_node_map = {}   # logical core to numa node id
```
I will add type info for both these list-of-list variables.
`node_*_cores` are list-of-list in sequence of node_core. i.e: `[[0, 1, 2, 3], [4, 5, 6, 7]]` for node 0 and 1 perspectively.
`*_core_node_map` are maps indicating with core is on which node. i.e.: `{0: 0, 1: 0, 2: 0, 3: 0, 4: 1, 5: 1, 6: 1, 7: 1}`

jingxu10 commented on torch/utils/launch_cpu_inference.py:
```diff
@@ -0,0 +1,577 @@
+import psutil
+from datetime import datetime
+
+format_str = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
```
the `get_logger()` doesn't work on my side.
I can get the output as following with command below:
```
format_str = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
logging.basicConfig(level=logging.INFO, format=format_str)
logger = logging.getLogger(__name__)
...

# torchrun --inference --nprocs 1 --node_id 1 --disable_iomp --no_python pwd
2022-04-26 05:36:44,250 - torch.utils.inference.run - INFO - Use JeMalloc memory allocator
2022-04-26 05:36:44,250 - torch.utils.inference.run - INFO - OMP_NUM_THREADS=XX
2022-04-26 05:36:44,250 - torch.utils.inference.run - INFO - LD_PRELOAD=/root/miniconda3/envs/py39/lib/libjemalloc.so
2022-04-26 05:36:44,250 - torch.utils.inference.run - INFO - numactl -C XX-XX -m 1 pwd
/workspace/srcs/cpu/pytorch
```

However, when I swith the logging to `get_logger()`, I cannot get logging messages:
```
from torch.distributed.elastic.utils.logging import get_logger
logger = get_logger()
...

# torchrun --inference --nprocs 1 --node_id 1 --disable_iomp --no_python pwd
/workspace/srcs/cpu/pytorch
```

kiukchung commented on test/test_torchrun.py:
```diff
@@ -0,0 +1,91 @@
+import subprocess
+
+class TestTorchrun(TestCase):
+    def test_jemalloc(self):
```
thanks for adding the test! could we move this to `test/utils/inference/run_test.py` to follow the test directory conventions as the rest of the project?

jingxu10 replied:
Sure.

kiukchung commented on test/test_torchrun.py:
```diff
@@ -0,0 +1,91 @@
+
+class TestTorchrun(TestCase):
+    def test_jemalloc(self):
+        if os.path.isdir('torchrun_logs'):
+            shutil.rmtree('torchrun_logs')
```
this can have unintended side-effects depending on the directory where the test is run from (you may end up deleting a legitimate `torchrun_logs` directory). Instead you can use the `setUp()` and `tearDown()` methods and create a temporary test directory for the test that is cleaned up after the test finishes running. Here's an example:

https://github.com/pytorch/pytorch/blob/master/test/distributed/elastic/agent/server/test/local_elastic_agent_test.py#L248-L253

jingxu10 replied:
done

kiukchung commented on test/test_torchrun.py:
```diff
@@ -0,0 +1,91 @@
+            shutil.rmtree('torchrun_logs')
+        num = 0
+        loc = os.path.dirname(os.path.abspath(__file__))
+        with subprocess.Popen("torchrun --inference --nprocs 1 --enable_jemalloc --log_path torchrun_logs --disable_numactl \
```
To keep the test assertions close to the logic that it is testing, I recommend testing `torch.utils.inference.run` as a module rather than through `torchrun --inference` CLI. 

For `torch.utils.run` you should definitely add a test that asserts that passing `--inference` indeed calls `torch.utils.inference.run` and passing nothing calls `torch.distributed.run`

jingxu10 replied:
done

kiukchung commented on docs/source/index.rst:
```diff
@@ -64,6 +64,7 @@ Features described in this documentation are classified by release status:
    torch.distributed.fsdp <fsdp>
    torch.distributed.optim <distributed.optim>
    torch.distributions <distributions>
+   utils/inference/run
```
lets follow the conventions here:

1. Create a main rst page for `docs/source/utils/run.rst` and give it the title `<torchrun>`. (e.g. `torchrun <utils.run>`)
2. In `utils/run.rst` explain at the high level what `torchrun` is doing, then link the two types of launchers from it (distributed.elastic.run and utils.inference.run)

kiukchung commented on docs/source/elastic/run.rst:
```diff
@@ -1,4 +1,4 @@
-.. _launcher-api:
+.. _launcher-ddp-api:
 
 torchrun (Elastic Launch)
```
this needs to change to `torch.distributed.run (Elastic Launch)` since we are making torchrun point to `torch.utils.run` now. You'll also have to edit:

1. https://github.com/pytorch/pytorch/blob/master/torch/distributed/run.py#L19
2.  https://github.com/pytorch/pytorch/blob/master/torch/distributed/launch.py#L7

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+
+class _Launcher():
+    r"""
+     Base class for launcher
```
looks like the documentation needs to be updated since `_Launcher` is no longer a base class (e.g. there are no abstract methods nor any subclasses of it)

jingxu10 replied:
Done.

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+        if platform.system() in ["Windows", "Darwin"]:
+            raise RuntimeError(f"{platform.system()} is not supported!!!")
+        elif platform.system() == "Linux":
+            args = ["lscpu", "--parse=CPU,Core,Socket,Node"]
+            lscpu_info = subprocess.check_output(args, universal_newlines=True).split("\n")
```
Lets add a comment here with a sample output of running lscpu, makes it easier to follow the rest of the logic. Something like

```
# Sample output of: `lscpu --parse=CPU,Core,Socket,Node`
# 
# # The following is the parsable format, which can be fed to other
# # programs. Each different item in every column has an unique ID
# # starting from zero.
# # CPU,Core,Socket,Node
# 0,0,0,0
# 1,1,0,0
...
```

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+# from torch.distributed.elastic.utils.logging import get_logger
+# logger = get_logger(__name__)
+
+class _CPUinfo():
```
would be good to add a unittest validating the correctness of the parsing logic in this class. You can do this by mocking the various return values of `lscpu --parse=CPU,Core,Socket,Node` and assert that `num_nodes, node_physical_cores, node_logical_cores, physical_core_node_map, logical_core_node_map` and the various getters indeed return the correct values.

jingxu10 replied:
done

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+            if numa_id not in numa_ids:
+                numa_ids.append(numa_id)
+        if len(numa_ids) > 1:
+            logger.warning(f"Numa Aware: cores:{str(core_list)} on different NUMA nodes:{str(numa_ids)}")
```
As part of the warning message, it would be great to add a short description (or even a link to a URL with the instructions) on how to fix this so that the next run does not display the warning again.

jingxu10 replied:
done

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+To get the peak performance on Intel(R) Xeon(R) Scalable Processors, the script optimizes the configuration of thread and memory
+management. For thread management, the script configures thread affinity and the preload of Intel OMP library.
+For memory management, it configures NUMA binding and preload optimized memory allocation library (e.g. tcmalloc, jemalloc).
+
```
could you also briefly discuss the auto omp settings and the various other environment variable this launcher auto-sets on behalf of the user, it could be as simple as a list of env vars that this launcher may set depending on the options and a brief description of what it does (a link to a more detailed doc is also acceptable)

jingxu10 replied:
done

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+*** Multi-instance inference ***
+
+1. Multi-instance
+   By default, one instance per node. if you want to set the instance numbers and core per instance,
```
`By default this tool runs one process per node`

jingxu10 replied:
Done. Thanks for the correction.

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+r"""
+This is a script for launching PyTorch inference on Intel(R) Xeon(R) Scalable Processors with optimal configurations.
+Single instance inference, multi-instance inference are enabled.
```
Could you add a description of what an "instance" is in this context (or switch to using "process"). For folks used to cloud, "instance" refers to a type of machine or node. So multi-instance implies multi-machine inference.

jingxu10 replied:
This script will be executed as a process, and will invoke multiple (`--ninstances` or `number of cores/--ncore_per_instance`) "instances". each "instance" will also be executed with multiple threads. "instance" here is kind of a group of multiple threads, not a real process. Probably adding description is a good idea.

jingxu10 replied:
To align with `--ncore_per_instance`, I think it might be better to use `--ninstances`. What do you think?

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+
+::
+
+   >>> torchrun --inference  --nprocs 14 --ncore_per_instance 4 python_script args
```
(see my comments above about procs vs instance), I'm guessing instance == proc, so for this setting, it is running 14 processes (aka instances) and 4 cores per process (so occupies 14 * 4 = 56 cores)

jingxu10 replied:
That's affirmative.

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+
+::
+
+   >>> torchrun --inference  --nprocs 2 --instance_idx 0 python_script args
```
we typically use `--rank` for this, any reason why this is called `--instance_idx`? In general I'd like to be consistent with terminology since there is already a lot going on under the hood.

jingxu10 replied:
Sure.

kiukchung commented on torch/utils/inference/run.py:
```diff
@@ -0,0 +1,569 @@
+r"""
+This is a script for launching PyTorch inference on Intel(R) Xeon(R) Scalable Processors with optimal configurations.
```
Correct me if I'm wrong here, but aren't the settings that are configured by this launcher also apply to training? I'm wondering why we are limiting these settings to inference only.

jingxu10 replied:
Yes, they should also work for training on CPUs. This time we focus on inference functionality. We plan to file a separate PR for training later.

kiukchung commented on torch/utils/run.py:
```diff
@@ -0,0 +1,24 @@
+
+    argv_script = sys.argv.copy()
+    del argv_script[0]
+    if mode == "--inference":
```
Sorry for not catching this earlier. It is preferable to use argparse.SubCommand instead of parsing the CLI here manually. See: https://docs.python.org/3/library/argparse.html#sub-commands

So this would be:

```
$ torchrun inference <.. args to torch.utils.inference.run...>
```

Using subcommand has many benefits like being able to print out help string for the main CLI as well as subcommands (e.g `torchrun --help` versus `torchrun inference --help`) and argparse takes care of mutual exclusivity of subcommands.

Here's an example of this in the TorchX CLI: https://github.com/pytorch/torchx/blob/main/torchx/cli/main.py#L29

There are various subcommands registered for `torchx` CLI (e.g. `$torchx run` `$torchx status` `$torchx log`)

jingxu10 replied:
Done

kiukchung reviewed (changes requested):
Thanks for making the edits! Looking better, left some more comments

malfet reviewed (changes requested):
I'd like to request that you change name of the runner to `torch.utils.run_cpu` or make it more generic, as were suggested in https://github.com/pytorch/pytorch/pull/63932#issuecomment-972623584

jingxu10 commented:
Hi @malfet , do you want me to rename `torch/utils/run.py` to `torch/utils/run_cpu.py` or `torch/utils/inference/run.py` to `torch/utils/inference/run_cpu.py`? The former one is a wrapper script that dispatches the path to ddp or inference. It is made to be the entrypoint of torchrun. I understand your request is for the later one, right?
By the way, I see @kiukchung had imported this PR already. Will my further edit on this PR bring side effects to him?

albanD commented on docs/source/conf.py:
```diff
@@ -113,7 +113,7 @@
     "config_from_args",
     "determine_local_world_size",
-    "get_args_parser",
+    "create_args",
```
No new entry should be added to this file in general.
Functions that are public API must be properly documented.

jingxu10 replied:
This is a list for coverage_ignore_functions, which are not expected to be exposed to users.

albanD replied:
Functions that should not be exposed should not be public. You can make that function private (by making it start with an _)  to avoid having to add it to this list.

jingxu10 replied:
Sure

albanD reviewed (changes requested):
Note that empty `__init__.py` files are not needed anymore.

albanD commented:
> By the way, I see kiukchung had imported this PR already. Will my further edit on this PR bring side effects to him?

No, he will only need to re-import the PR once it is updated.

jingxu10 commented:
> Note that empty `__init__.py` files are not needed anymore.

Hi, could you let me know which empty `__init__.py` is ok to be removed?

jingxu10 commented on torch/utils/run.py:
```diff
@@ -0,0 +1,26 @@
+from torch.distributed import run as run_ddp
+
+def main():
+    parser = argparse.ArgumentParser(description="torchrun CLI")
```
Yes, the entrypoint of `torchrun` is redirected to this script. Inside this script, distributed training will be dispatched to `torch.distributed.run`

jingxu10 replied:
> > By the way, I see kiukchung had imported this PR already. Will my further edit on this PR bring side effects to him?
> 
> No, he will only need to re-import the PR once it is updated.

Got it. Thanks.

albanD replied:
Ho sorry I missed the change to setup.py in this PR in the middle of all the rest.

jingxu10 commented on torch/utils/run.py:
```diff
@@ -0,0 +1,26 @@
+from torch.utils.inference import run as run_inf
+from torch.distributed import run as run_ddp
+
+def main():
```
As a discussion that we had with @kiukchung , inference launch script is also integrated into `torchrun`. The dispatching to inference path or ddp path is handled inside `torch/utils/run.py`, which is also the entrypoint of `torchrun`.

kiukchung replied:
created a thread for us on slack so that we can iterate quicker on discussions.

jingxu10 commented:
> > By the way, I see kiukchung had imported this PR already. Will my further edit on this PR bring side effects to him?
> 
> No, he will only need to re-import the PR once it is updated.

Got it. Thanks.

albanD commented:
The `torch.utils.inference.__init__.py`

jingxu10 commented:
> The `torch.utils.inference.__init__.py`

This empty __init__.py is to solve the following lint issue.
`Cannot find implementation or library stub for module named "torch.utils.inference"`.

Does it make sense to add the empty __init__.py back? @albanD

jingxu10 commented:
> > This empty init.py is to solve the following lint issue.
> > Cannot find implementation or library stub for module named "torch.utils.inference".
> > Does it make sense to add the empty init.py back? @albanD
> 
> Ho mypy is not using plain python import and is divergent. That's unfortunate. We can add it back then sure.

Got it. Done.

jingxu10 committed:
add launch script to torch.backends.xeon module

correct module name in run_cpu.py

update UT

update module name

add __init__.py to torch/backends/xeon

fix build doc error

add skipif to the UT to run on Linux only

import unittest in test_launch.sh

fix improper skipif in test_launch.py

albanD commented on test/backends/xeon/run_test.py:
```diff
@@ -0,0 +1,112 @@
+"""
+        from torch.backends.xeon.run_cpu import _CPUinfo
+        cpuinfo = _CPUinfo(lscpu_info)
+        assert cpuinfo._physical_core_nums() == 8
```
nit: use `self.assertEqual()` for these.

albanD commented on test/backends/xeon/run_test.py:
```diff
@@ -0,0 +1,112 @@
+        assert num == 4, "Failed to launch multiple instances for inference"
+
+if __name__ == "__main__":
+    subprocess.call("yes | conda install -c conda-forge gperftools jemalloc", shell=True)
```
I don't think we ever do such things in our tests.
I think it is better to raise a nice error if they are not available and hint the user towards installing these.

jingxu10 replied:
This command is to enable testing of preloading jemalloc and tcmalloc. If this cannot be executed, then I'll need to remove `test_jemalloc` (line 51) and `test_tcmalloc` (line 62) from this UT script. Is it OK to remove these 2 test functions?

albanD replied:
I think the way to go would be to check if these are installed and if not, either:
- Skip the tests that require it.
- Raise an error asking the user to install these before running the test.

albanD commented on test/backends/xeon/run_test.py:
```diff
@@ -0,0 +1,112 @@
+# Owner(s): ["module: unknown"]
```
Should this be run in CI?
If so:
- The file should be renamed to be `test_*` so that it actually runs
- We should have a proper module for it. Should I add a `module: xeon` and add you to the cc list for it?
- We need to make sure that it runs on a machine that has all the proper requirements. Do we have any CI machine with the right hardware to run this?

jingxu10 replied:
- Yeah, sure. I'll rename it to `test_launch_xeon.py`.
- Yeah, sounds good to me.
- 4 out of 5 test functions depend on installation of Jemalloc, TCMalloc and IOMP python packages (`test_jemalloc`, `test_tcmalloc`, `test_iomp` and `test_default_config`). If the installation cannot be done, I'll remove them, but just keep `test_multi_threads` to verify if multiple instances can be launched. This one doesn't require any hardware resources, just a loop to execute a bash command. Do you think if this one is needed? If not, I think we can remove this UT script.

albanD replied:
cc @malfet who is more familiar with the config on our current CI machines? Do we have the right hardware? And if so, can we install these libraries by default on the runners?

albanD commented on torch/backends/xeon/run_cpu.py:
```diff
@@ -0,0 +1,662 @@
+
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Is this needed? We only support python 3.7+

albanD commented on torch/backends/xeon/run_cpu.py:
```diff
@@ -0,0 +1,662 @@
+format_str = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+logging.basicConfig(level=logging.INFO, format=format_str)
+logger = logging.getLogger(__name__)
+# from torch.distributed.elastic.utils.logging import get_logger
```
spurious comment?

albanD commented on torch/backends/xeon/run_cpu.py:
```diff
@@ -0,0 +1,662 @@
+from typing import List, Dict
+
+format_str = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+logging.basicConfig(level=logging.INFO, format=format_str)
```
Should this be done inside the if main below?

albanD replied:
This file is imported during testing so this is going to change the global state of logging during testing?

albanD commented on torch/backends/xeon/run_cpu.py:
```diff
@@ -0,0 +1,662 @@
+                lscpu_cmd = ["lscpu", "--parse=CPU,Core,Socket,Node"]
+                lscpu_info = subprocess.check_output(lscpu_cmd, universal_newlines=True).split("\n")
+            else:
+                print("Running test to verify correctness of CPUinfo class")
```
log?

albanD reviewed (commented):
Thanks for the update on this!
I mainly have questions about the test but the rest looks good.

albanD commented on test/backends/xeon/test_launch.py:
```diff
@@ -0,0 +1,63 @@
+# Owner(s): ["module: unknown"]
```
This one should be `module: intel` ?

jingxu10 commented:
Hi @albanD , I cannot reply under your comment. Regarding the logging configuration, they (line 137 and 138) are not necessary if logging is configured somewhere else.

albanD reviewed (commented):
The update sounds good to me.
Can you please make sure the tests pass though? If that is only going to work on XEON machines, you can skip the test (with unittest.skipIf()) if you're not on XEON.

jingxu10 commented:
Hi @albanD , I've done fixing CI failures.
2 failing checks:
1. build docs. `torch.backends.xeon` is not added to *.rst files. I think you will update docs after the integration.
2. `usage_log.txt : The process cannot access the file because it is being used by another process.` It seems to be related to the CI environment.

albanD commented:
> build docs. torch.backends.xeon is not added to *.rst files. I think you will update docs after the integration.

This will need to be fixed before merging.
The `.rst` files in question are located in `docs/`
Depending on how you want to document this new submodule, you can update that.  You most likely just want to add it with the others in https://pytorch.org/docs/stable/backends.html?highlight=backends#module-torch.backends by modifying `backends.rst`.

albanD commented:
CI failure is real.

jingxu10 commented:
Oh, the launch script doesn't support Windows... I'll skip this UT on Windows.

albanD commented on test/backends/xeon/test_launch.py:
```diff
@@ -0,0 +1,66 @@
+    def tearDown(self):
+        shutil.rmtree(self._test_dir)
+
+    @unittest.skipIf(IS_LINUX, "Windows not supported")
```
- You can apply this to the class constructor on top if you want to skip all the tests in that class
- You want to skip if it is NOT linux right?
- it is both windows and macos that won't run? The comment could be "Only works on linux" if that's the case.

albanD reviewed (approved):
Thanks for the update.
Looks good to me.
Could you rebase on latest master to ensure that CI does run fine?

jingxu10 commented:
Sure. Will do. Thanks.

albanD commented:
@pytorchbot merge -g
